sentinel/src/schedule/reminders.js

```diff
@@ -1,225 +1,295 @@
+  if (place.tasks && place.tasks.find(task => task.form === reminder.form && task.timestamp === date.toISOString())) {
+    return false;
+  }
+
+  return true;
+};
+
+const filterValidPlaces = (reminder, date, placeIds) => {
+  if (!placeIds.length) {
+    return [];
+  }
+  return db.medic
+    .allDocs({ keys: placeIds, include_docs: true })
+    .then(result => {
+      const places = result.rows
+        .map(row => row.doc)
+        .filter(place => canSend(reminder, date, place));
+
+      if (!places.length) {
+        return [];
+      }
+
+      return lineage.hydrateDocs(places);
+    });
+};
+
+const getLeafPlaces = (reminder, date, startDocId) => {
+  return getLeafPlaceIds(startDocId).then(placeIds => {
+    // don't process the same doc twice!
+    placeIds = placeIds.filter(id => id !== startDocId);
+    if (!placeIds.length) {
+      return { places: [] };
+    }
+    const nextDocId = placeIds[placeIds.length - 1];
+    return filterPlaceIdsWithoutReminder(reminder, date, placeIds)
+      .then(placeIds => filterPlacesWithoutSentForms(reminder, date, placeIds))
+      .then(placeIds => filterValidPlaces(reminder, date, placeIds))
+      .then(places => ({ places, nextDocId }));
+  });
+};
+
+const createReminder = (reminder, date, place) => {
+  const context = {
+    templateContext: {
+      week: date.format('w'),
+      year: date.format('YYYY')
     },
-    _lineage: lineage
+    patient: place
+  };
+
+  const reminderDoc = {
+    _id: getReminderId(reminder, date, place._id ),
+    type: 'reminder',
+    contact: lineage.minifyLineage(place.contact),
+    place: { _id: place._id, parent: lineage.minifyLineage(place.parent) },
+    form: reminder.form,
+    reported_date: new Date().getTime(),
+    tasks: []
+  };
+  const task = messages.addMessage(reminderDoc, reminder, 'reporting_unit', context);
+  if (task) {
+    task.form = reminder.form;
+    task.timestamp = date.toISOString();
+    task.type = 'reminder';
+  }
+
+  return reminderDoc;
+};
+
+const sendReminders = (reminder, date, startDocId) => {
+  return getLeafPlaces(reminder, date, startDocId)
+    .then(({ places, nextDocId }) => {
+      if (!places || !places.length) {
+        return nextDocId;
+      }
+
+      const reminderDocs = places.map(place => createReminder(reminder, date, place));
+      return db.medic
+        .bulkDocs(reminderDocs)
+        .then(results => {
+          const errors = results.filter(result => result.error);
+          if (errors.length) {
+            logger.error('Errors saving reminders', errors);
+            throw new Error('Errors saving reminders');
+          }
+        })
+        .then(() => nextDocId);
+    })
+    .then(nextDocId => nextDocId && sendReminders(reminder, date, nextDocId));
+};
+
+const createReminderLog = (reminder, date, start) => {
+  const duration = performance.now() - start;
+  const reminderLog = {
+    _id: `reminderlog:${reminder.form}:${date.valueOf()}`,
+    reminder: reminder,
```

Comment:
Can this have a `type`?

Also, is this perhaps something that would be dealt with better with server-side telemetry?

That is to say, we temporarily store this level of detail and then aggregate it into a single doc once a month?

Interested in @garethbowen 's thoughts here.

Reply:
Added `type`. Maybe we should see how much noise this actually makes before we add another schedule. If schedules run once a week, 4 docs a month is little to clean up.